### ISSUE TITLE
Phrasal search

### ISSUE THREAD

**Issue by sylvinus** [commonsearch/cosr-front#18]
Queries like `"new york times"` should search for the exact sequence of words.

Elasticsearch [supports this](https://www.elastic.co/guide/en/elasticsearch/guide/current/phrase-matching.html), but it seems that if we want to mix phrase matches with regular matches we will have to do some preprocessing?

**Comment by wumpus** [comment]
Do you really mean "new york times" in quotes, or the query [new york times] with no quotes? (I use the contention to put query text in [] to avoid ambiguity... then quotes really mean quotes.)

**Comment by sylvinus** [comment]
Right! I added the Markdown formatting but let's try to stick with [] for queries indeed! I meant `["new york times"]` :)

**Comment by wumpus** [comment]
OK with the quotes, isn't it trivial to do the necessary query parsing?

The non-quotes case is also interesting, because it's much more common. But we can talk about that another time!

**Comment by sylvinus** [comment]
You're right - this probably isn't a `hard` issue, I'll downgrade it :)

To make the issue clearer, we also need to support queries like `["new york times" 2016]`

**Comment by amirouche** [comment]
FWIW it's called _phrasal search_

**Comment by sylvinus** [comment]
Thanks @amirouche :) Want to have a swing at it ? ;-)

**Comment by amirouche** [comment]
> ;-)

So, you read my profile and you know that I am working on [a similar project using Scheme](https://github.com/amirouche/culturia) and hence I mostly do Scheme nowdays. I was just lurking around. Technically our projects differs in two important ways 1) you have multiple repositories for every component. Culturia has only one. I want to xp with single repository approach used at big5 companys. 2) In GNU Guile (GG!) there is no GIL (ahah!) hence viable multiple threads and a very single processus. I want it to be very easy to sysadmin also. For instance, in commonsearch work done by elasticsearch is done on a custom search engine built on top of wiredtiger storage engine. One of the advantage is that when I store my data in the FTS index I can span the ACID transaction until the correct data is written in the graph database part of the same storage engine. Otherwise said it's like ACID txn between PostgreSQL and Neo4j (which is difficult to do!). Which allows to share in a coherent set the actual "search data" and the crawling data. This could allows to implement some crazy transderivational search. And they will always be in sync. This is kind of parano thing but I like it. The bad news is scaling over multiple nodes is terrible. Its search speed is ok. CommonSearch is damn fast! What is the size of the index?

> Elasticsearch supports this, but it seems that if we want to mix phrase matches with regular matches we will have to do some preprocessing

What do you mean by preprocessing step? Do you want to change the way documents are indexed in ES? Or it's about crafting the correct ES json incantation? Also does the Query parser recognize the double quote operation as phrasal search somehow. For you culture (!), in culturia (!), it's implemented using an inverted index `(term-id, position-id, doc-id)`. E.g. to do phrasale search of `["new work times"]`, the search engine will first do `[new]` search and retrieve a set of `(term-id, position-id, doc-id)` (it might be a generator over the results), then it only have to intersect with documents which follow the pattern `("work", position-id+1, {{ doc-id }})` and then intersect with `("times", position-id+2, {{doc-id}})`.

By the way I am `amz3` on #python-fr come and chat.
